### PR TITLE
simulate pid and motion magic

### PR DESCRIPTION
**Pull request by ariovistus**
at least make it available for simulations to call..

### CODE REVIEW

File: tests/run_tests.sh [hunk 1]
```diff
@@ -3,8 +3,7 @@
 cd $(dirname $0)
 
 if [ "$RUNCOVERAGE" == "1" ]; then
-    python3 -m coverage run --source ctre -m pytest "$@"
-    python3 -m coverage report -m
+    python3 -m pytest --cov-report html --cov ctre "$@"
```
Comment (auscompgeek) [inline tests/run_tests.sh]: I'm all for using pytest-cov, but you need to:
* make the report readable from Travis CI, so let it report to the terminal, and
* install pytest-cov on Travis CI.

File: gen/MotController_data.yml [hunk 2]
```diff
@@ -120,19 +121,32 @@ c_MotController_Create1:
...
+      'pid1_outBeforRmp': 0,
+      'pid1_target': 0,
       
       # slotIdx
       'profile_slot_select': 0,
       # pidIdx
       'pid_slot_select': 0,
       
+      # motion magic parameters
+      'cruise_velocity': 0,
```
Comment (virtuald) [inline gen/MotController_data.yml]: This all looks fantastic, but I'm going to bikeshed here... shouldn't cruise_velocity and acceleration have magic or mm or something to indicate that they're part of motion magic?

File: gen/MotController_data.yml [hunk 3]
```diff
@@ -90,7 +91,7 @@ c_MotController_Create1:
       
       # These two values are set when the set()/setDemand() functions are called
       'control_mode': 0,
-      'value': 0,
+      'value': 0, # percent throttle
```
Comment (virtuald) [inline gen/MotController_data.yml]: Not your bug, but this has to be 0.0 or the get function will fail.

File: tests/test_talonsrx.py [hunk 4]
```diff
@@ -485,15 +463,181 @@ def test_basemotorcontroller_selectedSensorPosition(talon, cdata):
...
+    assert talon.getClosedLoopTarget(0) == 44
+    assert talon.getClosedLoopTarget(1) == 55
+
+
 def test_basemotorcontroller_valueUpdated(talon):
     talon.valueUpdated()
+
+
+def assert_is_almost_equal(expected, actual, rel_tol=0, abs_tol=0.00001):
+    assert math.isclose(expected, actual, rel_tol=rel_tol, abs_tol=abs_tol), "expected: %f, got: %f" % (expected, actual)
```
Comment (auscompgeek) [inline tests/test_talonsrx.py]: I still think this assert should be moved to each test function. pytest will do introspection for you, so no need for the message, or the mess of specifying your own `rel_tol` and `abs_tol` that don't match what you actually want.

File: gen/MotController_data.yml [hunk 5]
```diff
@@ -120,23 +122,40 @@ c_MotController_Create1:
...
+      'motionmagic_cruise_velocity': 0,
+      'motionmagic_acceleration': 0,
+      'motionmagic_target': 0,
+      'motionmagic_velocity': 0,
       
     })
   
     hal_data['CAN'][deviceId] = self.hal_data
+
+    if self.Notifier:
```
Comment (virtuald) [inline gen/MotController_data.yml]: Shouldn't this only run when magic/pid mode is enabled?

Comment (ariovistus) [reply gen/MotController_data.yml]: _calculate_1ms won't do anything unless its in magic/pid mode. I suppose you could enable/disable in setDemand.

Comment (virtuald) [reply gen/MotController_data.yml]: That would be ideal; no sense creating more threads if they aren't needed.

Comment (ariovistus) [reply gen/MotController_data.yml]: not at a place where I can check easily, is there a free function somewhere in the talon class? that notifier is probably going to have to be stopped at some point

Comment (virtuald) [reply gen/MotController_data.yml]: The notifier has a free function and will stop itself.

**Pull request merged**